### DISCUSSION TITLE
Taking config file from a branch other than repository default

### DISCUSSION THREAD

**Discussion by tarquin-the-brave** [renovatebot/renovate#13036] [Suggest an Idea]
The docs state that the Renovate config can be placed in a bunch of file locations on the repository's default branch. https://docs.renovatebot.com/configuration-options/

This is a bit awkward.  I'm currently trialling Renovate for use in a Gitlab codebase.  I've been very impressed with how non-intrusive it is (up to this point).  I've been able to set up a job to run Renovate and I was impressed by its ability to autodetect dependencies.

I would like to fully proof of concept the use of Renovate on a branch, seeing what the dependency dashboard looks like, etc, but find myself needing to merge the config I want to use into the default branch.  This means I can't change the config file without getting a merge request approved and going through a review process.

Maybe I've missed a CLI parameter to say "use the config file at this local path", but if I haven't, that would be a really useful feature to have.

**Comment by viceice** [comment]
nope, config needs to be on default branch, use a fork to test 😉

**Reply by TWiStErRob** [reply]
How do you fork a repository you own?

**Reply by szpak** [reply]
Just create a separate group/organization, e.g. TWiStErRob-forks.

**Comment by rarkins** [comment]
You can do a PoC without needing a config in any branch:
- Put all config in your `config.js` which you pass to the bot
- Include `requireConfig: false` and `onboarding: false`
- Renovate will then proceed without onboarding, and create PRs targeting the default branch

Alternatively if you want to commit a config to some other branch (e.g. `foobranch`):
- Configure `baseBranches: ['foobranch']` in your bot config
- Add a `renovate.json` to `foobranch` manually
- Renovate will then proceed without onboarding, and create PRs targeting `fooobranch`

**Reply by tarquin-the-brave** [reply]
Thanks.

From reading the docs https://docs.renovatebot.com/configuration-options/:

> Renovate always uses the config from the repository's default branch, even if that configuration specifies multiple baseBranches. Renovate does not read/override the config from within each base branch if present.

I read this as meaning that even if I specify my branch as a `baseBranch` it will still look for config on the default branch.

**Reply by viceice** [reply]
I think you are interested in 
- https://github.com/renovatebot/renovate/pull/12514
- https://github.com/renovatebot/renovate/issues/2496

**Reply by tarquin-the-brave** [reply]
I don't see how this 2nd option is supposed to work.  I've got renovate running on a pipeline in my branch, and it is not picking up the config file on the same branch.  How do I get the `baseBranches: ['foobranch']` config to Renovate?  without it being on the default branch?  I can't see any CLI parameter for that.

**Reply by viceice** [reply]
then  you only option is to follow the way `rarkins` wrote.

**Reply by TWiStErRob** [reply]
I used `@rarkins`'s first solution ([here's the code](https://github.com/TWiStErRob/renovate-config/tree/33e91ed5f3dead265d4ac39459c66d6d5998fa1b)), and it mostly worked, except I had to put my config file in `force` block.

**Reply by davinkevin** [reply]
I tried your solution @rarkins, and result is still attached to `main`, not the `foobranch` I used. 

This really interesting solution you proposed is still valid?

**Reply by rarkins** [reply]
I haven't tested it in sometime. Usually it's required to have a config in the default branch unless you set requireConfig=optional

**Reply by davinkevin** [reply]
Even with it, I had to commit something to `main` unfortunately. 

I really like renovate, but a way to work without any commit would be very good to reduce time to first result

**Reply by rarkins** [reply]
With requireConfig=optional you don't need any config/commit

**Reply by davinkevin** [reply]
Hello @rarkins, 

I think I found why this behaves differently for me. Let me describe my configuration: 

We have: 
* a project with `main` as the GitHub default branch
* Our updates target the branch named `release`, and that one hosts the configuration we want to use, so we have `"useBaseBranchConfig": "merge"` and `"baseBranches": ["release"]` on `main`

In that context, I'm not able to follow steps you describe for the second part of your answer ("Alternatively if you want to commit a config to some other branch") because I can't override, from the bot config, the `baseBranches` parameter. I think all those `useBaseBranchConfig` and `baseBranches` cause the trouble related to my issue. 

My goal is simple, I would like to open a PR to my project (commit, dedicated feature branch `foobranch`), but for testing purpose, I would like to run `renovate` bot locally (using docker or else) like if that branch was the official configuration

If you want, I can open a PR for that and we can plan a pairing session if you need to see it for real 😇.

**Reply by rarkins** [reply]
Thanks, please start a new discussion laying out your requirements first, then proposed solution

**Reply by davinkevin** [reply]
[Done](https://github.com/renovatebot/renovate/discussions/31603)

**Reply by Rockdell** [reply]
Hi! First of all sorry for the wall of text 🙏 
I couldn't get this to work when targeting a self-hosted Azure DevOps Server. Even with the settings suggested Renovate does not recognize the renovate.json file on the defined baseBranch. Even with "requireConfig" and "onboarding" set to false and "useBaseBranchConfig" set to "merge" it gives out this error message: `"message": "Cannot read properties of null (reading 'ignorePresets')",`. Which I imagine is because these is no config.js file on the defaultBranch and it is trying to get the value from a non-existent config, but that's just my guess.

**Suggestion:** It would be awesome if you could just skip the check for the defaultBranch config file and use the renovate.json file (or other config files) on the baseBranches I want to target.

Initially this was propose to be implemented here #2496, with the "replace" option for the "useBaseBranchConfig" setting but then it changed into the "merge" option we have today. This is useful in cases where a main branch is not used, for example, in my company we have a branch for each released version of our product with some versions receiving updates, and needing to have a config file on a default branch that is not used or that will need to be updated at some point is cumbersome.

I know I can set "requireConfig" to "ignored" and have the entire config someplace else but it would be nice to have the option to override some settings on some specific branches if desired.

**Reply by jamietanna** [reply]
@Rockdell please can you raise another Discussion? We try not to bump old threads

**Comment by klangenk** [comment]
You can use this configuration now
https://docs.renovatebot.com/configuration-options/#usebasebranchconfig

**Comment by tnelson-doghouse** [comment]
useBaseBranchConfig doesn't work for me.  I'm wanting to set base branch patterns in renovate-config.js (or environment variables), which sets baseBranchPatterns to a specific branch

The options are currently:
- merge: This reads the config from the base branch, which overrides the other things, then doesn't read the branch I actually wanted because it's just overridden the baseBranchPatterns.  
- none: This also reads config from the base branch, and then ignores the branch I wanted

What we need is a "skip" option, that skips the base branch for the repo, and just goes straight to the named branch.  

And since our organisation is set up around a workflow that doesn't really support forks, that's not an option either.

**Reply by jamietanna** [reply]
@tnelson-doghouse please can you raise another Discussion? We try not to bump old threads

**Comment by jamietanna** [comment]
Closing as `useBaseBranchConfig` solved the original ask - fresh requests can be raised in a new Discussion - thanks!
